### PR TITLE
Roadrunner as simulator

### PR DESCRIPTION
**Pull request by PaulJonasJost**
Added functionality for RoadRunner simulations. Currently still dependent on PEtab (which might be removed at a later stage). Unifying PEtabImporter for amici, roadrunner and general scenarios will be put into another PR to have this one standalone.

Tests for roadrunner are petab_testsuite except for 3 tests that expectedly fail. One is just not supported in pyPESTO (Timepoint-specific parameter overrides), one is still not supported in functionality (more complex noise formulas) and one is not clear whether this is an intended test (proclaiming a species as parameter in .xml).

### CODE REVIEW

**Review by dweindl — COMMENTED** [review]
Cool! 

It think filling some gaps in validation before merging would be great. 

> Unifying PEtabImporter for amici, roadrunner and general scenarios will be put into another PR to have this one standalone.

I'd do that only for the parts where it's really easily possible. PEtab handling is already sufficiently complex for a single simulator.

> Tests for roadrunner are petab_testsuite except for 3 tests that expectedly fail. One is just not supported in pyPESTO (Timepoint-specific parameter overrides), 

For the current gradient-free setup, this would be easy to support. For other cases, see also [petab.flatten_timepoint_specific_output_overrides](https://petab.readthedocs.io/projects/libpetab-python/en/latest/modules.html#petab.flatten_timepoint_specific_output_overrides). Not a priority for me, though.


> one is not clear whether this is an intended test (proclaiming a species as parameter in .xml)

Intended, as explained elsewhere.

File: pypesto/objective/roadrunner/petab_importer_roadrunner.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/petab_importer_roadrunner.py:105]: The idea was the following:
Petablint has already been run, therefore the petab problem is "well posed" and any parameter that occurs in the observable table has not been in the sbml model. Thus if it now appears in the roadrunner model it may be due to the following two things:
* Someone manually added these parameters
* Someone already ran the `_write_observables_to_model`

I would say in both cases they should be notified of this but the creation of the parameters is fine, and the assignment rules for observables will get overwritten.

Does this reasoning make sense? Or why specifically would you raise an error here?

File: pypesto/objective/roadrunner/petab_importer_roadrunner.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/petab_importer_roadrunner.py:121]: See above discussion.

File: pypesto/objective/roadrunner/petab_importer_roadrunner.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/petab_importer_roadrunner.py:123]: :+1:

File: pypesto/objective/roadrunner/petab_importer_roadrunner.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/petab_importer_roadrunner.py:255]: great idea, thanks :)

File: pypesto/objective/roadrunner/road_runner.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/road_runner.py:75]: Changed to "Check whether we need some sort of pickling"

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/roadrunner_calculator.py:1]: Not sure what you mean. There is a "module-level" docstring for the roadrunner module which is similar to the other objective function modules. Or do you mean within each file?

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/roadrunner_calculator.py:18]: good idea, changed to measurement/simulation/noise

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/roadrunner_calculator.py:88]: :+1:

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/roadrunner_calculator.py:126]: :+1:

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/roadrunner_calculator.py:145]: It is because of how roadrunner is constructed. As it has flexible meaning to time, the first value is the one where roadrunner starts its simulation (therefore we need 0 as the first value) and it also means it needs a second value to know up till where it should simulate

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/roadrunner_calculator.py:249]: :+1:

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/roadrunner_calculator.py:357]: i just naively assumed it was french and used it :D yeah counterpart is better :+1:

File: pypesto/objective/roadrunner/utils.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/utils.py:39]: it is, but i think having the option to retrieve it via self.timepoints is convenient, and it also gives the option for a check whether they actually coincide.

File: pypesto/objective/roadrunner/utils.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/utils.py:62]: agreed, currently is serves as a placeholder till i know what kind of sanity checks i want to run.

File: pypesto/objective/roadrunner/utils.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/utils.py:143]: added a better message.

File: pypesto/objective/roadrunner/petab_importer_roadrunner.py
Comment (dweindl) [reply pypesto/objective/roadrunner/petab_importer_roadrunner.py:105]: > I would say in both cases they should be notified 

I wouldn't want to support either of those cases. If the model was modified externally between validation and calling this function, it's in an undefined state and a lot of things can go wrong.

File: pypesto/objective/roadrunner/utils.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/utils.py:404]: :+1:

File: pypesto/objective/roadrunner/utils.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/utils.py:447]: :+1:

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (dweindl) [reply pypesto/objective/roadrunner/roadrunner_calculator.py:1]: Every Python file is a module. pypesto.objective.roadrunner is a subpackage. Right, I meant each file.

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (dweindl) [reply pypesto/objective/roadrunner/roadrunner_calculator.py:145]: So `[0.0]` will not suffice to get the initial state?

File: pypesto/objective/roadrunner/utils.py
Comment (dweindl) [reply pypesto/objective/roadrunner/utils.py:39]: Fine to add a function/property for that (i.e. returning the first column), but don't duplicate the data.

> it also gives the option for a check whether they actually coincide

If you don't duplicate it, you don't have this option, but you have the guarantee that they don't differ :)

File: test/base/test_roadrunner.py
Comment (dweindl) [inline test/base/test_roadrunner.py:103]:
```suggestion
            f"Case {case}: Test results do not match expectations"
```

File: test/base/test_roadrunner.py
Comment (dweindl) [inline test/base/test_roadrunner.py:93]:
```suggestion
        f"LLH: simulated: {llh}, expected: {gt_llh}, match = {llhs_match}",
```

File: pypesto/objective/roadrunner/__init__.py
Comment (dweindl) [inline pypesto/objective/roadrunner/__init__.py:1]: Please add this to `doc/api.rst`

**Review by dweindl — COMMENTED** [review]
Please add the new files/dir to `.github/CODEOWNERS`.

File: pypesto/objective/roadrunner/utils.py
Comment (dweindl) [inline pypesto/objective/roadrunner/utils.py:166]: The time has come: https://github.com/sys-bio/roadrunner/releases/tag/v2.6.0

File: pypesto/objective/roadrunner/petab_importer_roadrunner.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/petab_importer_roadrunner.py:108]: ok, just removed the try..except. Makes sense to call an error here I guess with your reasoning. 👍🏼

File: pypesto/objective/roadrunner/petab_importer_roadrunner.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/petab_importer_roadrunner.py:168]: true. removed

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/roadrunner_calculator.py:1]: phuu yeah will do, will probably need to apply this then to all pypesto files as well 🙈

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/roadrunner_calculator.py:156]: no, [0.0] is not really possible. Should I catch that here? 🤔

Comment (dweindl) [reply pypesto/objective/roadrunner/roadrunner_calculator.py:156]: Ok. No, no change required.

File: pypesto/objective/roadrunner/utils.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/utils.py:294]: yes, tried and replaced it now 👍🏼  will keep the upscale_parameter**s**, as it is independent of the petab.problem.Problem

File: pypesto/objective/roadrunner/utils.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/utils.py:166]: yes I know :D thanks :)

**Comment by PaulJonasJost** [comment]
> Please add the new files/dir to `.github/CODEOWNERS`.

Not really needed imo as I am sole code owner of objective? 🤔

File: pypesto/objective/roadrunner/utils.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/utils.py]: integrated 👍🏼

File: .github/workflows/deploy.yml
Comment (dweindl) [inline .github/workflows/deploy.yml:12]: Either token, or trusted publishing don't mix.
Revert `.github/workflows/deploy.yml` here in favor of #1341.

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (dweindl) [inline pypesto/objective/roadrunner/roadrunner_calculator.py:3]:
```suggestion
Handles all RoadRunner.simulate calls, calculates likelihoods and residuals.
```

**Review by dweindl — APPROVED** [review]
LGTM

File: .github/workflows/deploy.yml
Comment (PaulJonasJost) [reply .github/workflows/deploy.yml]: reverted as it has its own separate PR 👍🏼

**Comment by dweindl** [comment]
```
  /home/runner/work/pyPESTO/pyPESTO/pypesto/objective/roadrunner/utils.py:304: SettingWithCopyWarning: 
  A value is trying to be set on a copy of a slice from a DataFrame.
  Try using .loc[row_indexer,col_indexer] = value instead
```

**Comment by PaulJonasJost** [comment]
> ```
>   /home/runner/work/pyPESTO/pyPESTO/pypesto/objective/roadrunner/utils.py:304: SettingWithCopyWarning: 
>   A value is trying to be set on a copy of a slice from a DataFrame.
>   Try using .loc[row_indexer,col_indexer] = value instead
> ```

from what in understand on [pandas website](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy) this will go away anyways at pandas 3.0

**Comment by dweindl** [comment]
> this will go away anyways at pandas 3.0

Well, the warning will go away, yes. It will be an error by then.

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (Doresic) [inline pypesto/objective/roadrunner/roadrunner_calculator.py:112]: Use pypesto constant FVAL? 
And make constants for the other two, if they don't exist already?

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (Doresic) [inline pypesto/objective/roadrunner/roadrunner_calculator.py:127]: Same as for FVAL

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (Doresic) [inline pypesto/objective/roadrunner/roadrunner_calculator.py:274]: It is a bit tough to understand what the different entries of the parameter mapping are. Would it be better if `parameter_mapping` was a dictionary, with keys describing what kind of objects it contains? Similar to amici's `map_sim_var`, `map_sim_var_preeq` and so on.

Not crucial as this is really a under the hood kind of a thing, but would be easier to understand the code.

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (Doresic) [inline pypesto/objective/roadrunner/roadrunner_calculator.py:398]: Should there be a warning here? First trying to see if there are more or less nan values in the llh. And then in the first case warning that there are additional nan values in the llh calculation, or the opposite? (Not sure if the second is possible, but first can probably occur)

**Review by Doresic — APPROVED** [review]
Great implementation. Thanks!

I didn't try out the specific simulation and measurement structures you construct, so I don't know if it can be more efficient, but it looks good as it is. You might be able to make the llh evaluations faster by combining the matrices of observables with the same noise_distribution (i.e. same llh type). That would reduce the number of for-loop steps. But would be uglier in the sense of implementation.

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/roadrunner_calculator.py:398]: I am not sure what you mean 🤔  if we encounter a nan-value in llh, there are two options: in the first case, it is also an nan value in the measurements - that's fine, we don't really care, as it indicates that there was no measurement to begin with, in the other case it is not, and independent of who has more nan values, this should result in a nan value of the objective, as the simulation corresponding to a valid measurement was not possible. That was at least my thought process here. Happy for suggestions :)

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (PaulJonasJost) [reply pypesto/objective/roadrunner/roadrunner_calculator.py]: added it into the doctring description and added comments. Think that should suffice for better code understanding for now.

File: pypesto/objective/roadrunner/roadrunner_calculator.py
Comment (Doresic) [reply pypesto/objective/roadrunner/roadrunner_calculator.py:411]: True, my mistake. I misunderstood a part of it, but still have a question.
I was concerned with the case when the llh is nan where a measurement is not nan. This can happen if the simulation is nan for that specific non-nan measurement. In that case, it's the job of the simulator to warn that there are nan values.

But I'm asking if there is a case when both simulation and measurement are non-nan, but the llh is nan. In this case, would you want to pass a warning here that you saw a nan value?

Definitely not important in the big picture and ok as it is.

**Pull request merged**
